File: src/week2/report2/TicketOffice.java
package week2.report2;

import java.util.HashMap;
import java.util.Map;

public class TicketOffice {
  private Money amount;
  private Map<Theater, Double> commissionRate = new HashMap<>();

  public TicketOffice(Money amount) {
    this.amount = amount;
  }

  boolean contract(Theater theater, Double rate) {
    if (commissionRate.containsKey(theater))
      return false;
    commissionRate.put(theater, rate);
    return true;
  }

  boolean cancel(Theater theater) {
    if (!commissionRate.containsKey(theater))
      return false;
    commissionRate.remove(theater);
    return true;
  }

  Reservation reserve(Theater theater, Movie movie, Auditorium auditorium, Screening screening, int count) {
    if (!commissionRate.containsKey(theater) || !theater.isValidScreening(movie, auditorium, screening)
        || !auditorium.hasSeat(screening, count))
      return Reservation.NONE;
    Reservation reservation = theater.reserve(movie, auditorium, screening, count);
    if (reservation != Reservation.NONE) {
      Money sales = movie.calculateFee(screening, count);
      Money commission = sales.multi(commissionRate.get(theater));
      amount = amount.plus(commission);
      theater.plusAmount(sales.minus(commission));
    }
    return reservation;
  }
}

File: Swift/week2/Week2/Week2/Example/TicketOffice.swift
//
//  TicketOffice.swift
//  Week2
//
//  Created by tskim on 18/10/2019.
//  Copyright © 2019 assin. All rights reserved.
//

import Foundation


class TicketOffice: MemoryHashable {
    private var amount: Money
    private var commissionRate = [Theater: Double]()
            
    init(amount: Money) {
        self.amount = amount
    }
    
    func contract(theater: Theater, rate: Double) -> Bool {
        
        if commissionRate.containKey(key: theater) {
            return false
        }
        commissionRate.updateValue(rate, forKey: theater)
        return true
    }
    func cancel(theater: Theater) -> Bool {
        
        if !commissionRate.containKey(key: theater) {
            return false
        }
        commissionRate.removeValue(forKey: theater)
        return true
    }
    
    func reserve(theater: Theater, movie: Movie, screening: Screening, count: Int) -> Reservation {
        if !commissionRate.containKey(key: theater) || !theater.isValidScreening(movie: movie, screening: screening) || !screening.hasSeat(count: count) {
            return Reservation.None
        }
        
        let reservation = theater.reserve(movie: movie, screening: screening, count: count)
        if reservation !== Reservation.None {
            
            let sales = movie.calculateFee(screening: screening, audienceCount: count)
            let commission = sales.multi(times: commissionRate[theater] ?? 1)
            self.amount = self.amount.plus(amount: commission)
            theater.plusAmount(amount: sales.minus(amount: commission))
        }
        return reservation
    }
}
